### PR TITLE
Update the input props type

### PR DESCRIPTION
**Pull request by higorquintao**
If applied, this pull request will update the input props type

### Other minor changes:
 - no minor changes

### Card Link:
https://github.com/codelittinc/agnostic-design-system/issues/484

### Notes:
- It was created a new generic type to be reused for all inputs and selectors.

### CODE REVIEW

File: src/components/Dropdown/renderOptions.tsx [hunk 1]
```diff
@@ -88,7 +89,7 @@ const RenderOptions = <T extends {}>(props: DropdownProps<T>, ref?: React.Ref<HT
             : styles['dropdown-list-positioned-bottom']
         )}
         listItemCategory={listItemCategory}
-        selected={selected}
+        selected={value}
```
Comment (paulohfev) [inline src/components/Dropdown/renderOptions.tsx:92]: @higorquintao we're getting a TS error from this line. It seems to be a missing property type warning. Can you please check?

File: src/components/Dropdown/renderOptions.tsx [hunk 2]
```diff
@@ -123,7 +124,7 @@ const RenderOptions = <T extends {}>(props: DropdownProps<T>, ref?: React.Ref<HT
           nodeBeforeItems={!hideGroupByTitle && buildAppendList(item, index)}
           onChange={onChange}
           options={groupedOptions[item]}
-          selected={selected}
+          selected={value}
```
Comment (paulohfev) [inline src/components/Dropdown/renderOptions.tsx:127]: @higorquintao we're getting a TS error from this line. It seems to be a missing property type warning. Can you please check?

File: src/components/Input/index.tsx [hunk 3]
```diff
@@ -19,7 +19,13 @@ export type Filter = {
   test: RegExp | Function;
 };
 
-export interface Props {
+export type ControlledInputProps<T> = {
+  value?: T;
+  onChange: (value?: T, event?: any) => void;
+  onStateChange?: (state: boolean) => void;
```
Comment (victor0402) [inline src/components/Input/index.tsx:25]: @higorquintao I see you're using this callback to communicate whether the input is valid or not, but when reading "onStateChange" I would expect to receive the actual input state as the argument, and its value. 

Can we change the name of the function? 
A suggestion would be `onValidationChange?: (isValid: boolean)`

File: src/components/Input/index.tsx [hunk 4]
```diff
@@ -213,7 +220,7 @@ const Input = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
         value={getAttr(value || '')}
         onFocus={getAttr(() => {
           setValidationState('');
-          onFocus();
+          onFocus && onFocus();
```
Comment (victor0402) [inline src/components/Input/index.tsx:223]:
```suggestion
      onFocus?.();
```

File: src/components/List/index.tsx [hunk 5]
```diff
@@ -5,6 +5,9 @@ import { isArray, isEqual } from 'lodash';
 import styles from '@/components/List/List.css';
 
 import ListItem from '@/components/ListItem';
+/* eslint-disable */
+import { ControlledInputProps } from '../Input';
```
Comment (victor0402) [inline src/components/List/index.tsx:9]: why eslint is complaining here?

File: src/components/Dropdown/index.tsx [hunk 6]
```diff
@@ -61,7 +59,7 @@ const Dropdown = <T extends {}>(props: Props<T>) => {
 
   const handleValidation = (item?: T | T[]) => {
     const valid = !required || !isEmpty(item);
-    onStateChange(valid);
+    onStateChange && onStateChange(valid);
```
Comment (victor0402) [inline src/components/Dropdown/index.tsx:62]:
```suggestion
    onStateChange?.(valid);
```

File: src/components/Dropdown/InputDropdown/index.tsx [hunk 7]
```diff
@@ -140,9 +138,9 @@ const InputDropdown = <T extends {}>(props: Props<T>) => {
     }
   };
 
-  const handleChangeInput = e => {
-    onInputChange && onInputChange(e);
-    setListTitle(e.target.value);
+  const handleChangeInput = (value, event) => {
+    onInputChange && onInputChange(event);
```
Comment (victor0402) [inline src/components/Dropdown/InputDropdown/index.tsx:142]:
```suggestion
    onInputChange?.(event);
```

**Review by victor0402 — CHANGES_REQUESTED** [review]

**Review by paulohfev — CHANGES_REQUESTED** [review]
please check out the requested changes on the List component

**Pull request merged**